---
title: "Chip-seq in Galaxy and R"
author: "Emily Chambers"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output: 
  html_notebook: 
    toc: yes
    toc_float: yes
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,message=FALSE)
```
# Hands-on chip-seq Analysis in Galaxy and R

### Sheffield Bioinformatics Core
<img src="images/logo-sm.png" align=right>

web : [sbc.shef.ac.uk](http://sbc.shef.ac.uk)  
twitter: [SheffBioinfCore](https://twitter.com/SheffBioinfCore)  
email: [bioinformatics-core@sheffield.ac.uk](bioinformatics-core@sheffield.ac.uk)

-----

# Acknowledgement

Some of these materials are based on courses available at .

- [ATAC-Seq data analysis in Galaxy](https://training.galaxyproject.org/training-material/topics/epigenetics/tutorials/atac-seq/tutorial.html/)
- [Chip-seq excersizes in R by Jonas Ibn-Salem, JGU Mainz](http://cbdm-01.zdv.uni-mainz.de/~jibnsale/teaching/chip-seq_exercises.html)
- [The ChIPpeakAnno user’s guide by Jianhong Ou, Jun Yu, Lihua Julie Zhu](https://www.bioconductor.org/packages/release/bioc/vignettes/ChIPpeakAnno/inst/doc/ChIPpeakAnno.html)

-----

# Tutorial overview

This tutorial will cover  the analysis of the ChIP-seq peak data aquired in part 1.

Lets re-examine our workflow. We now have our peaks called in a bed file and are ready for our downstram analysis.

![](images/park_workflow.jpg)

---

# 1 Requirements
Install the packages we will need for the analysis

```{r}
if(!require(ChIPpeakAnno)) BiocManager::install("ChIPpeakAnno")
if(!require(TxDb.Hsapiens.UCSC.hg18.knownGene)) BiocManager::install("TxDb.Hsapiens.UCSC.hg18.knownGene")
if(!require(org.Hs.eg.db))BiocManager::install("org.Hs.eg.db")
if(!require(dplyr)) BiocManager::install("dplyr")
if(!require(rtracklayer)) BiocManager::install("rtracklayer")
if(!require(reactome.db)) BiocManager::install("reactome.db")

```

# 2. Upload Data

Upload our MACS2 peak bed files that we downloaded from galaxy

```{r}
library(readr)
FOXA1_df <- read_tsv("data/FOXA1_full.bed", col_names = T, comment = "#"  )
FOXA1_df

# bed format
ER <- import("data/ER_full.bed", format="bed")
```

# 3. Convert peaks into a GRanges object
A GRanges object form the GenomicRanges packages is a container for genomic locations and their associated annotations. We will use this data structure for the ChIP-seq peak coordinates.
```{r}
library(GenomicRanges) # load the GenomicRanges pakcage

FOXA1 <- GRanges(
  FOXA1_df$chr,
  IRanges(FOXA1_df$start, FOXA1_df$end),
  strand="*"
)

# we can add more data to each peak subsequently
names(FOXA1) <- paste("FOXA1_peak", 1:nrow(FOXA1_df), sep="_")
score(FOXA1) <- FOXA1_df[,"-10*log10(pvalue)"]  

# show the first and last lines of the GRanges object
FOXA1
```

###   Excersie: Read in the ER peaks 

```{r, }
#We fix the name and score columns.

# assign scores by converting the 'name' feald to type numeric
score(ER) <- as.numeric(ER$name)

# overwrite the name column
ER$name <- paste("ER_peaks", 1:length(ER), sep="_")

# use the names() function 
names(ER) <- ER$name 
ER
```


# 4. Basic number and statistics of the FOXA1 peaks


What is the number of peaks?
Use the function length().

```{r}
length(FOXA1)
```
What is the mean, median, and max size of the peaks?
The function width() returns the sizes of all ranges in a GRanges object as vector.

```{r}
FOXA1.size <- width(FOXA1)
summary(FOXA1.size)
```

3.4 What is the distribution of peak sizes?
Plot the distribution of sizes using the hist() function.


```{r}
hist(width(FOXA1), xlab="FOXA1 peak size", col="gray")
```


What is the distribution of FOXA1 ChIP-seq peak p-values?
We plot the histogram of −log10 transformed p-values. Remember that we stored them in the score column of the GRanges object. We can access them using the function score().

```{r}
# get the -log_10 transformed p-values from the score column
mlPvals <- score(FOXA1)
hist(mlPvals, xlab="-log_10(p-value)", col="gray")
```

# 5. Compare the peaks of ER and FOXA1

Make a barplot of the number of peaks
```{r}
bp <- barplot(c(length(ER), length(FOXA1)), names=c("ER", "FOXA1"))
# add actual values as text lables to the plot
text(bp, c(length(ER), length(FOXA1)), labels=c(length(ER), length(FOXA1)), pos=1)
```

Use handy function from `ChIPpeakAnno` to calculate the overlaps Then we use the package Vennerable to build a Venn object with the number of peaks in each subset. This object can than be passed to the ‘plot’ function.



## How many ER peaks overlap FOXA1 peaks?
```{r}

# find overlap of binding sites 
ovlHits <- findOverlapsOfPeaks(ER, FOXA1)
ovlHits$venn_cnt
overlaps <- ovlHits$peaklist[["ER///FOXA1"]]

```



```{r}
makeVennDiagram(list(ER, FOXA1), NameOfPeaks=c("ER", "FOXA1"),
                scaled=FALSE, 
                fill=c("#009E73", "#F0E442")) # circle fill color
```

A pie chart is used to demonstrate the overlap features of the common peaks.


```{r}

pie1(table(ovlHits $overlappingPeaks[["ER///FOXA1"]]$overlapFeature))
```



# 6. Functional annotation of ChIP-seq peaks
To understand the function of a transcription factor we want to know to which genes and genomic features it binds.
We will use a TxDb objects. Such an object is an R interface to prefabricated databases contained by specific annotation packages. The package TxDb.Hsapiens.UCSC.hg18.knownGene includes all human genes and transcripts from UCSC with coordinates for the hg18 genome assembly.

```{r}
library(TxDb.Hsapiens.UCSC.hg18.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg18.knownGene # just a shortcut

## create annotation file from EnsDb or TxDb
annoData <- toGRanges(txdb, feature="gene")
annoData[1:2]
rtracklayer::export(annoData, con="TxDb.Hsapiens.UCSC.hg18.gene.bed")
```

Now we overlap the annotation to the peaks
```{r}
overlaps_anno <- annotatePeakInBatch(overlaps, AnnotationData=annoData, 
                                    output="overlapping", maxgap=5000L)

head(overlaps_anno)

aCR<-assignChromosomeRegion(overlaps_anno, nucleotideLevel=FALSE, 
                           precedence=c("Promoters", "immediateDownstream", 
                                         "fiveUTRs", "threeUTRs", 
                                         "Exons", "Introns"), 
                           TxDb=TxDb.Hsapiens.UCSC.hg18.knownGene)
barplot(aCR$percentage)

```

Add the gene names to the annotation and write out as `csv` and `bed` files.

```{r}
library(dplyr)
library(org.Hs.eg.db)
overlaps_anno <- addGeneIDs(overlaps_anno, org.Hs.eg.db, IDs2Add = c("symbol","refseq", "ensembl"), 
feature_id_type = "entrez_id")



as.data.frame(unname(overlaps_anno)) %>% 
  dplyr::select(-peakNames,-peak) %>% 
  readr::write_csv("overlap_peaks_annotated.csv")

names(overlaps_anno) <- lapply(overlaps_anno$peakNames, function(x) paste(x,collapse=":"))
names(overlaps_anno) <- paste(overlaps_anno$feature,names(overlaps_anno),sep="_")


rtracklayer::export(overlaps_anno, con="overlap_peaks.bed")

```

Use functionality within ChIPpeakAnno to perform enrichment (see package vignette).

```{r}
over <- getEnrichedGO(overlaps_anno, orgAnn="org.Hs.eg.db",
                      maxP = 0.05, minGOterm = 10,
                      multiAdjMethod = "BH",condense = TRUE)
write_csv(over[["bp"]], path = "overlap_peaks_GO.csv")
```

## Enrichment with Reactome

```{r}
library(reactome.db)
path <- getEnrichedPATH(overlaps_anno, "org.Hs.eg.db","reactome.db", maxP = 0.05) %>% 
  rename(ENTREZID = EntrezID)

gene_anno <- AnnotationDbi::select(org.Hs.eg.db, keys = as.character(path$ENTREZID), keytype="ENTREZID",columns="SYMBOL")

path_anno <- AnnotationDbi::select(reactome.db, keys = as.character(path$path.id), 
                                   keytype="PATHID",
                                   columns=c("PATHID","PATHNAME")) %>% rename(path.id = PATHID)

path %>% left_join(gene_anno) %>% 
  group_by(path.id) %>% 
  mutate(Genes = paste(unique(SYMBOL),collapse=";")) %>% 
  ungroup %>% 
  filter(!duplicated(path.id)) %>% 
  left_join(path_anno) %>% 
  distinct %>% 
  select(-ENTREZID, -PATH, -totaltermInDataset, -totaltermInGenome, -SYMBOL)  %>% 
  arrange(pvalue) %>% 
  write_csv("overlap_peaks_reactome.csv")
```


