---
title: "Chip-seq in Galaxy and R"
author: "Emily Chambers"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output: 
  html_notebook: 
    toc: yes
    toc_float: yes
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,message=FALSE)
```
# Hands-on chip-seq Analysis in Galaxy and R

### Sheffield Bioinformatics Core
<img src="images/logo-sm.png" align=right>

web : [sbc.shef.ac.uk](http://sbc.shef.ac.uk)  
twitter: [SheffBioinfCore](https://twitter.com/SheffBioinfCore)  
email: [bioinformatics-core@sheffield.ac.uk](bioinformatics-core@sheffield.ac.uk)

-----

# Acknowledgement

Some of these materials are based on courses available at .

- [ATAC-Seq data analysis in Galaxy](https://training.galaxyproject.org/training-material/topics/epigenetics/tutorials/atac-seq/tutorial.html/)
- [Chip-seq excersizes in R by Jonas Ibn-Salem, JGU Mainz](http://cbdm-01.zdv.uni-mainz.de/~jibnsale/teaching/chip-seq_exercises.html)

-----

# Tutorial overview

This tutorial will cover  the analysis of the ChIP-seq peak data aquired in part 1.

Lets re-examine our workflow. We now have our peaks called in a bed file and are ready for our downstram analysis.

![](images/park_workflow.jpg)

---

# 1 Requirements
Install the package "ChIPpeakAnno" from Bioconductor:

```{r}
if(!require(ChIPpeakAnno)) BiocManager::install("ChIPpeakAnno")
if(!require(Vennerable)) install.packages("Vennrable")

```

# 2. Upload Data

Upload our MACS2 peak bed files that we downloaded from galaxy

```{r}
library(readr)
FOXA1_df <- read_tsv("data/FOXA1.bed", col_names = F)
FOXA1_df

ER_df <- read_tsv("data/ER.bed", col_names = F)
ER_df
```

# 3. Convert peaks into a GRanges object
A GRanges object form the GenomicRanges packages is a container for genomic locations and their associated annotations. We will use this data structure for the ChIP-seq peak coordinates.
```{r}
library(GenomicRanges) # load the GenomicRanges pakcage

FOXA1 <- GRanges(
  FOXA1_df$X1,
  IRanges(FOXA1_df$X2, FOXA1_df$X3),
  strand="*"
)

# we can add more data to each peak subsequently
names(FOXA1) <- paste("FOXA1_peak", 1:nrow(FOXA1_df), sep="_")
score(FOXA1) <- FOXA1_df$X8
  
# show the first and last lines of the GRanges object
FOXA1
```

```{r}
ER <- GRanges(
  ER_df$X1,
  IRanges(ER_df$X2, ER_df$X3),
  strand="*"
)

# we can add more data to each peak subsequently
names(ER) <- paste("ER_peak", 1:nrow(ER_df), sep="_")
score(ER) <- ER_df$X8
  
# show the first and last lines of the GRanges object
ER
```


# 4. Basic number and statistics of the FOXA1 peaks


What is the number of peaks?
Use the function length().

```{r}
length(FOXA1)
```
What is the mean, median, and max size of the peaks?
The function width() returns the sizes of all ranges in a GRanges object as vector.

```{r}
FOXA1.size <- width(FOXA1)
summary(FOXA1.size)
```

3.4 What is the distribution of peak sizes?
Plot the distribution of sizes using the hist() function.


```{r}
hist(width(FOXA1), xlab="FOXA1 peak size", col="gray")
```


What is the distribution of FOXA1 ChIP-seq peak p-values?
We plot the histogram of −log10 transformed p-values. Remember that we stored them in the score column of the GRanges object. We can access them using the function score().

```{r}
# get the -log_10 transformed p-values from the score column
mlPvals <- score(FOXA1)
hist(mlPvals, xlab="-log_10(p-value)", col="gray")
```

# 5. Compare the peaks of ER and FOXA1

Make a barplot of the number of peaks
```{r}
bp <- barplot(c(length(ER), length(FOXA1)), names=c("ER", "FOXA1"))
# add actual values as text lables to the plot
text(bp, c(length(ER), length(FOXA1)), labels=c(length(ER), length(FOXA1)), pos=1)
```

Use handy function from `ChIPpeakAnno` to calculate the overlaps Then we use the package Vennerable to build a Venn object with the number of peaks in each subset. This object can than be passed to the ‘plot’ function.

```{r}
# get subsets of binding sites
ER.uniq <- setdiff(ER, FOXA1)
FOXA1.uniq <- setdiff(FOXA1, ER)


# plot overlap of binding sites as Venn diagram


# build objects with the numbers of sites in the subsets
venn <- Venn(SetNames=c("ER", "FOXA1"), 
    Weight=c(
        '10'=length(ER.uniq), 
        '11'=length(ovl), 
        '01'=length(FOXA1.uniq)
    )
)

# plot Venn Diagram
plot(venn)
```

and make a venn diagram

```{r}
makeVennDiagram(ovl)
```


# 6. Functional annotation of ChIP-seq peaks
To understand the function of a transcription factor we want to know to which genes and genomic features it binds.

We will use a TxDb objects. Such an object is an R interface to prefabricated databases contained by specific annotation packages. The package TxDb.Hsapiens.UCSC.hg18.knownGene includes all human genes and transcripts from UCSC with coordinates for the hg18 genome assembly.

```{r}



# load gene annotation from UCSC for human genome build hg18
require(TxDb.Hsapiens.UCSC.hg18.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg18.knownGene # just a shortcut

```
